File: backend/app/services/room_service.py
from sqlalchemy.orm import Session
from sqlalchemy import func, and_, or_
from typing import List, Optional, Any, Dict
from uuid import UUID
from app.models.infrastructure import Room, RoomType, RoomStatus, Bed, BedStatus
from app.services.base import BaseService
from app.schemas.infrastructure import RoomCreate, RoomUpdate, RoomTypeCreate, RoomTypeUpdate

class RoomTypeService(BaseService[RoomType, RoomTypeCreate, RoomTypeUpdate]):
    def get_by_name(self, db: Session, name: str) -> Optional[RoomType]:
        return db.query(RoomType).filter(RoomType.name == name).first()

class RoomService(BaseService[Room, RoomCreate, RoomUpdate]):
    def create(self, db: Session, *, obj_in: RoomCreate) -> Room:
        # 1. Create the room itself
        room = super().create(db, obj_in=obj_in)
        
        # 2. Determine capacity (beds count)
        capacity = 0
        if room.room_type_id:
            room_type = db.query(RoomType).get(room.room_type_id)
            if room_type:
                capacity = room_type.capacity
        
        if capacity > 0:
            beds = []
            for i in range(1, capacity + 1):
                bed_name = f"G-{i}"
                bed = Bed(
                    label=bed_name,
                    room_id=room.id,
                    status=BedStatus.AVAILABLE
                )
                beds.append(bed)
            
            db.add_all(beds)
            db.commit()
            db.refresh(room)
            
        return room

    def _compute_occupancy(self, room: Room):
        if room and room.beds:
            room.current_occupancy = sum(1 for b in room.beds if b.status in [BedStatus.OCCUPIED, BedStatus.RESERVED])

    def get(self, db: Session, id: UUID) -> Optional[Room]:
        from sqlalchemy.orm import joinedload
        room = db.query(Room).options(
            joinedload(Room.beds),
            joinedload(Room.room_type),
            joinedload(Room.building)
        ).filter(Room.id == id).first()
        if room:
            self._compute_occupancy(room)
        return room

    def get_multi(self, db: Session, *, skip: int = 0, limit: int = 100) -> List[Room]:
        from sqlalchemy.orm import joinedload
        rooms = db.query(Room).options(
            joinedload(Room.beds),
            joinedload(Room.room_type),
            joinedload(Room.building)
        ).offset(skip).limit(limit).all()
        for room in rooms:
            self._compute_occupancy(room)
        return rooms

    def get_by_building(self, db: Session, building_code: str) -> List[Room]:
        from sqlalchemy.orm import joinedload
        rooms = db.query(Room).options(
            joinedload(Room.beds),
            joinedload(Room.room_type),
            joinedload(Room.building)
        ).filter(Room.code.startswith(building_code)).all()
        for room in rooms:
            self._compute_occupancy(room)
        return rooms

    def get_available_rooms(self, db: Session) -> List[Room]:
        from sqlalchemy.orm import joinedload
        rooms = db.query(Room).options(
            joinedload(Room.beds),
            joinedload(Room.room_type),
            joinedload(Room.building)
        ).filter(Room.status == RoomStatus.AVAILABLE).all()
        for room in rooms:
            self._compute_occupancy(room)
        return rooms

    def check_room_availability(self, db: Session, room_id: UUID) -> bool:
        room = self.get(db, room_id)
        if not room:
            return False
            
        capacity = 0
        if room.room_type:
            capacity = room.room_type.capacity
        else:
            capacity = len(room.beds)
        
        return room.current_occupancy < capacity

    def search_rooms(self, db: Session, 
                     room_type_id: Optional[UUID] = None, 
                     min_price: Optional[float] = None, 
                     max_price: Optional[float] = None,
                     status: Optional[RoomStatus] = None,
                     attributes: Optional[Dict[str, Any]] = None,
                     building_id: Optional[UUID] = None,
                     keyword: Optional[str] = None
                     ) -> List[Room]:
        from sqlalchemy.orm import joinedload
        query = db.query(Room).options(
            joinedload(Room.beds),
            joinedload(Room.room_type),
            joinedload(Room.building)
        )
        
        if room_type_id:
            query = query.filter(Room.room_type_id == room_type_id)
            
        if min_price is not None:
             query = query.join(RoomType, isouter=True)
             query = query.filter(func.coalesce(Room.base_price, RoomType.base_price) >= min_price)

        if max_price is not None:
             if not room_type_id: 
                query = query.join(RoomType, isouter=True)
             query = query.filter(func.coalesce(Room.base_price, RoomType.base_price) <= max_price)

        if status:
            query = query.filter(Room.status == status)

        if attributes:
            for key, value in attributes.items():
                query = query.filter(Room.attributes[key].astext == str(value))
        
        if building_id:
            query = query.filter(Room.building_id == building_id)

        if keyword:
            search = f"%{keyword}%"
            query = query.join(Room.building)
            query = query.filter(
                (Room.code.ilike(search)) 
            )

        rooms = query.all()
        for room in rooms:
            self._compute_occupancy(room)
        return rooms

    def get_occupancy_stats(self, db: Session, building_id: Optional[UUID] = None) -> Dict[str, Any]:
        from sqlalchemy.orm import joinedload
        query = db.query(Room).options(
            joinedload(Room.room_type), 
            joinedload(Room.beds),
            joinedload(Room.building)
        )
        if building_id:
            query = query.filter(Room.building_id == building_id)
        
        rooms = query.all()
        total_rooms = len(rooms)
        
        status_counts = {}
        total_capacity = 0
        total_occupied = 0
        
        building_stats_map = {}
        
        for room in rooms:
            s = room.status.value
            status_counts[s] = status_counts.get(s, 0) + 1
            
            cap = room.room_type.capacity if room.room_type else len(room.beds)
            total_capacity += cap
            
            occ = sum(1 for bed in room.beds if bed.status in [BedStatus.OCCUPIED, BedStatus.RESERVED])
            total_occupied += occ
            
            b_name = room.building.name if room.building else "Unknown"
            if b_name not in building_stats_map:
                building_stats_map[b_name] = {"name": b_name, "total": 0, "occupied": 0}
            
            building_stats_map[b_name]["total"] += cap
            building_stats_map[b_name]["occupied"] += occ

        return {
            "total_rooms": total_rooms,
            "status_breakdown": status_counts,
            "total_capacity": total_capacity,
            "total_occupied": total_occupied,
            "occupancy_rate": (total_occupied / total_capacity * 100) if total_capacity > 0 else 0,
            "building_stats": list(building_stats_map.values())
        }

room_type_service = RoomTypeService(RoomType)
room_service = RoomService(Room)

File: frontend/src/services/room-service.ts
import { api } from './api';

export interface Room {
    id: string; // UUID
    code: string;
    floor: number;
    // building is now an object
    building?: {
        id: string;
        name: string;
        code: string;
        campus_id?: string;
        campus?: {
            id: string;
            name: string;
        };
    };
    building_id: string; // Needed for forms
    // room_type is now an object
    room_type?: {
        id: string;
        name: string;
        capacity: number;
        base_price: number;
        description?: string;
    };
    room_type_id: string; // Needed for forms
    capacity: number;
    current_occupancy: number;
    status: 'AVAILABLE' | 'FULL' | 'MAINTENANCE' | 'RESERVED';
    base_price: number;
    description?: string;
    beds: { id: string; label: string; status: string }[];
    gender_type: string;
    area_m2?: number;
    attributes?: Record<string, any>;
}

export interface RoomCreate {
    code: string;
    floor: number;
    building_id: string;
    room_type_id: string;
    status?: string;
    gender_type: string;
    base_price: number;
    area_m2?: number;
    attributes?: Record<string, any>;
}

export interface RoomUpdate {
    code?: string;
    floor?: number;
    building_id?: string;
    room_type_id?: string;
    status?: string;
    gender_allowed?: string;
    area_m2?: number;
    attributes?: Record<string, any>;
}

export interface RoomType {
    id: string;
    name: string;
    description?: string;
    capacity: number;
    base_price: number;
}

export const roomTypeService = {
    async getRoomTypes(): Promise<RoomType[]> {
        const response = await api.get<RoomType[]>('/api/v1/rooms/types');
        return response.data;
    },

    async createRoomType(data: any): Promise<RoomType> {
        const response = await api.post<RoomType>('/api/v1/rooms/types', data);
        return response.data;
    },

    async updateRoomType(id: string, data: any): Promise<RoomType> {
        const response = await api.put<RoomType>(`/api/v1/rooms/types/${id}`, data);
        return response.data;
    },

    async deleteRoomType(id: string): Promise<RoomType> {
        const response = await api.delete<RoomType>(`/api/v1/rooms/types/${id}`);
        return response.data;
    }
};

export const roomService = {
    roomType: roomTypeService,

    async getRooms(params?: { building?: string; status?: string; skip?: number; limit?: number; search?: string }): Promise<Room[]> {
        // Backend expects 'keyword' for search
        const queryParams: any = { ...params };
        if (params?.search) {
            queryParams.keyword = params.search;
            delete queryParams.search;
        }
        const response = await api.get<Room[]>('/api/v1/rooms/', { params: queryParams });
        return response.data;
    },

    async getRoom(id: string): Promise<Room> {
        const response = await api.get<Room>(`/api/v1/rooms/${id}`);
        return response.data;
    },

    async createRoom(data: RoomCreate): Promise<Room> {
        const response = await api.post<Room>('/api/v1/rooms/', data);
        return response.data;
    },

    async updateRoom(id: string, data: RoomUpdate): Promise<Room> {
        const response = await api.put<Room>(`/api/v1/rooms/${id}`, data);
        return response.data;
    },

    async deleteRoom(id: string): Promise<void> {
        await api.delete(`/api/v1/rooms/${id}`);
    },

    async getCampuses(): Promise<{ id: string; name: string }[]> {
        const response = await api.get('/api/v1/rooms/campuses');
        return response.data;
    },

    async getBuildings(campus_id?: string): Promise<{ id: string; name: string; code: string }[]> {
        const response = await api.get('/api/v1/rooms/buildings', { params: { campus_id } });
        return response.data;
    },

    async getRoomStats(building_id?: string): Promise<any> {
        const response = await api.get('/api/v1/rooms/stats', { params: { building_id } });
        return response.data;
    }
};
